source/assets/js/playground/console-utils.ts

```diff
@@ -45,36 +59,92 @@ function lineNumberFormatter(number?: number): string {
   return `${number}`;
 }
 
-export function displayForConsoleLog(item: ConsoleLog): string {
-  const data: {type: string; lineNumber?: number; message: string} = {
+function selectionLink(
+  playgroundState: PlaygroundState,
+  range?: PlaygroundState['selection']
+): string {
+  if (!range) return '';
+  return serializeState({...playgroundState, selection: range});
+}
+
+export function displayForConsoleLog(
+  item: ConsoleLog,
+  playgroundState: PlaygroundState
+): string {
+  const data: {
+    type: string;
+    lineNumber?: number;
+    message: string;
+    safeLink?: string;
+    range?: PlaygroundState['selection'];
+  } = {
     type: item.type,
     lineNumber: undefined,
     message: '',
+    safeLink: undefined,
+    range: null,
   };
   if (item.type === 'error') {
     if (item.error instanceof Exception) {
-      data.lineNumber = item.error.span.start.line;
+      const span = item.error.span;
+      data.lineNumber = span.start.line;
+      data.range = [
+        span.start.line + 1,
+        span.start.column + 1,
+        span.end.line + 1,
+        span.end.column + 1,
+      ];
     }
     data.message = item.error?.toString() || '';
   } else if (['debug', 'warn'].includes(item.type)) {
     data.message = item.message;
     let lineNumber = item.options.span?.start?.line;
+    if (item.options.span) {
+      const span = item.options.span;
+      data.lineNumber = span.start.line;
+      data.range = [
+        span.start.line + 1,
+        span.start.column + 1,
+        span.end.line + 1,
+        span.end.column + 1,
+      ];
+    }
+
     if (typeof lineNumber === 'undefined') {
       const stack = 'stack' in item.options ? item.options.stack : '';
-      const needleFromStackRegex = /^- (\d+):/;
+      const needleFromStackRegex = /^- (\d+):(\d+) /;
       const match = stack?.match(needleFromStackRegex);
       if (match?.[1]) {
         // Stack trace starts at 1, all others come from span, which starts at
         // 0, so adjust before formatting.
         lineNumber = parseInt(match[1]) - 1;
       }
+      if (match?.[2]) {
+        data.range = [
+          parseInt(match[1]),
+          parseInt(match[2]),
+          parseInt(match[1]),
+          parseInt(match[2]),
+        ];
+      }
     }
     data.lineNumber = lineNumber;
+
+    if (item.type === 'warn' && item.options.deprecationType?.id) {
+      data.safeLink = `https://sass-lang.com/d/${item.options.deprecationType.id}`;
```

Comment:
I find the indirection in this safeLink thing a bit confusing, and I worry it'll get worse if we end up wanting to add more HTML to the messages. Consider instead having separate `htmlMessage` and `message` variables and injecting `htmlMessage ?? encodeHTML(message)`.

Reply:
This wouldn't work in this situation, unless I misunderstand. I'll explain what I'm doing here a bit more to see if there's a clearer and more future-proof method.

1. `message` comes from the Logger, and is inherently unsafe
2. When we have a deprecationType, we can generate a known safe URL for more info
3. If that known safe URL exists as a string in `message`, we want to turn that string into a link. 

I realized the link augmentation is a bit confusing to include in the htmlEncode function, so I moved it out in  https://github.com/sass/sass-site/pull/1179/commits/e86f9b9fe5a5926f8fcceb7930dbf43886a67400.